src/System.Threading.Timer/tests/TimerFiringTests.cs

```diff
@@ -0,0 +1,226 @@
+            Assert.False(are.WaitOne(20), "Should not have received a second timer event");
+            t.Change(1, -1);
+            Assert.True(are.WaitOne(50), "Should have received a second timer event after changing it");
+        }
+    }
+
+    [Fact, OuterLoop]
+    public void Running_Timer_CanBeDisposedAndStopsFiring()
+    {
+        AutoResetEvent are = new AutoResetEvent(false);
+        TimerCallback tc = new TimerCallback((object o) => are.Set());
+        var t = new Timer(tc, null, 100, 500);
+        Assert.True(are.WaitOne(250), "Failed to get first timer fire");
+        t = null; // Remove our refence so the timer can be GC'd 
+        GC.Collect();
+        Assert.False(are.WaitOne(750), "Should not have received a timer fire after it was collected");
+    }
+
+    [Fact, OuterLoop]
+    public void MultpleTimers_PeriodicTimerIsntBlockedByABlockedTimer()
+    {
+        object t1State = new object();
+        object t2State = new object();
+        AutoResetEvent areT1 = new AutoResetEvent(false);
+        AutoResetEvent areT2 = new AutoResetEvent(false);
+        TimerCallback tc = new TimerCallback((object o) =>
+        {
+            if (o == t1State)
+            {
+                areT1.Set();
+            }
+            else
+            {
+                areT2.Set();
+            }
+        });
+
+        using (var t1 = new Timer(tc, t1State, 500, -1))
+        using (var t2 = new Timer(tc, t2State, 100, 100))
+        {
+            Assert.False(areT1.WaitOne(0), "Should not have a signaled t1ARE");
+            Assert.True(areT2.WaitOne(150), "Should have received a periodic timer fire");
+            Assert.True(areT2.WaitOne(150), "Should have received a second periodic timer fire");
+            Assert.True(areT1.WaitOne(500), "Should have received the blocked t1 timer fire");
+        }
+    }
+
+    [Fact, OuterLoop]
+    public void ManyTimers_EachTimerDoesFire()
+    {
+        int maxTimers = 10000;
+        object syncLock = new object();
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+        TimerCallback tc = new TimerCallback((object o) =>
+        {
+            lock (syncLock)
+            {
+                count++;
+                if (count >= maxTimers)
+                    are.Set();
+            }
+        });
+
+        Timer[] arr = new Timer[maxTimers];
+        try
+        {
+            for (int i = 0; i < maxTimers; i++)
+            {
+                arr[i] = new Timer(tc, null, 20 /* enough time to wait on the ARE */, -1);
+            }
+
+            bool success = are.WaitOne(250);
+            lock (syncLock)
+            {
+                Assert.True(success, String.Format("Not all timers fired: only {0} of {1}", count, maxTimers));
+            }
+        }
+        finally
+        {
+            foreach (Timer t in arr)
+                if (t != null)
+                    t.Dispose();
+        }
+    }
```

Comment:
This could be simplified using a CountdownEvent, e.g.
```C#
public void ManyTimers_EachTimerDoesFire()
{
    const int NumTimers = 10000;
    CountdownEvent ce = new CountdownEvent(NumTimers);
    Timer[] timers = Enumerable.Range(0, 10000).Select(_ => new Timer(s => ce.Signal(), null, 20, -1))).ToArray();
     try { Assert.True(ce.Wait(250)); }
     finally { foreach (Timer t in timers) t.Dispose(); }
}
```
or something like that.